src/uct/tcp/tcp_sockcm.c

```diff
@@ -36,18 +37,73 @@ static uct_cm_ops_t uct_tcp_sockcm_ops = {
     .listener_create  = UCS_CLASS_NEW_FUNC_NAME(uct_tcp_listener_t),
     .listener_reject  = uct_tcp_listener_reject,
     .listener_query   = uct_tcp_listener_query,
-    .listener_destroy = UCS_CLASS_DELETE_FUNC_NAME(uct_tcp_listener_t)
+    .listener_destroy = UCS_CLASS_DELETE_FUNC_NAME(uct_tcp_listener_t),
+    .ep_create        = UCS_CLASS_NEW_FUNC_NAME(uct_tcp_sockcm_ep_t)
 };
 
+static void uct_tcp_close_sa_arg(uct_tcp_sa_arg_t *sa_arg_ctx)
+{
+    ucs_list_del(&sa_arg_ctx->list);
+    ucs_async_remove_handler(sa_arg_ctx->fd, 1);
+    close(sa_arg_ctx->fd);
+    ucs_free(sa_arg_ctx);
+}
+
+static int uct_tcp_is_fd_connected(int fd)
+{
+    ucs_status_t status;
+    int ret;
+
+    ret = ucs_socket_is_connected(fd);
+    if (!ret) {
+        ucs_error("connection establishment for fd %d was unsuccessful", fd);
+
+        status = ucs_async_modify_handler(fd, 0);
+        if (status != UCS_OK) {
+            ucs_error("failed to modify %d event handler to 0: %s",
+                      fd, ucs_status_string(status));
+        }
+    }
+
+    return ret;
+}
+
 void uct_tcp_sa_data_handler(int fd, void *arg)
 {
-    ucs_warn("Function not implemented");
+    uct_tcp_sa_arg_t *sa_arg_ctx = (uct_tcp_sa_arg_t *)arg;
+    uct_tcp_sockcm_ep_t *cep     = sa_arg_ctx->ep;
```

Comment:
`cep` -> `ep` or `cm_ep`? or "c" means smth?

Reply:
"c" means "connected" EP, we use this notation in rdmacm impl